programs/survival/compass.sc

```diff
@@ -0,0 +1,611 @@
+                // Remove the obsidian without causing updates, then the game handles breaking an air block, then put the obsidian back
+                without_updates(
+                    set(pos, 'air');
+                    schedule(0, _(outer(pos)) -> (set(pos, 'obsidian')));
+                );
+            );
+        );
+    );
+);
+
+track(toTrack) -> (
+    player = player();
+
+    slotNum = player ~ 'selected_slot';
+
+    slot = player ~ 'holds';
+
+    if (slot: 0 != 'compass',
+        // If the player isn't holding a compass, say so
+        print(player, format('lb '+slot: 0, 'l s can\'t track things'));
+    ,
+        if (!(slot: 2 ~ 'Tracking'),
+            // If the compass is a regular compass, say so
+            print(player, format('l This is just a regular compass... ', 'c [Get another compass]', '! /compass giveCompass'));
+        ,
+            // If logOnTrack is true, tell the player who's being tracked that they're being tracked
+            if (global_logOnTrack,
+                print(player(toTrack), format('lb '+player, 'l  is tracking you!'));
+            );
+
+            // Change the player that the compass is tracking
+            slot: 2: 'Tracking' = toTrack;  
+            inventory_set(player, slotNum, slot: 1, slot: 0, slot: 2);
+            print(format('l Tracking ', 'lb '+toTrack));
+        );
+    );
+);
+
+giveCompass(player) -> (
+    canGive = global_maxCompasses == 'infinity';
+
+    if (!canGive,
+        // If maxCompasses is a number, find the amount of compasses the player has and check if it's less than the maximum amount
+        slot = inventory_find(player, 'compass');
+        slots = 0;
+        while (slot != null, 41,
+            slots = slots + inventory_get(player, slot): 1;
+            slot = inventory_find(player, 'compass', slot+1);
+        );
+
+        canGive = slots < global_maxCompasses;
+    );
+
+    if (canGive,
+        // Find all the players that aren't the player receiving the compass
+        players = filter(entity_selector('@a'),
+            _ != player;
+        );
+
+        if (length(players) > 0, 
+            if (global_logOnTrack,
+                print(player(players: 0), format('lb '+player, 'l  is tracking you!'));
+            );
+
+            nbtData = '{Tracking:"'+lower(players: 0 ~ 'name')+'",LodestoneTracked:false,LodestonePos:{X:0,Y:0,Z:0},LodestoneDimension:"overworld",display:{Name:\'{"text":"'+players: if(i, i, 0) ~ 'name'+'"}\'}}';
+            
+            run('give '+player ~ 'name'+' minecraft:compass'+nbtData+' 1');
+        ,
+            // If there are no other players online, say so
+            print(player, format('l There are no players online'));
+        );
+    ,
+        print(player, format('l You already have the maximum amount of compasses!'));
+    );
+);
+
+item_consumed(slot) -> (
```

Comment:
just curious about it - why did you use signals to trigger item_consumed function rather than calling it directly from your app code?

Reply:
The idea is that other scripts that want to consume items can signal that event to make the persist option work with their own scripts

Reply:
then shouldn't it be added to some sort of library function maybe?

Reply:
I think this ought to be added to some form of library function tbh which could be built-in, which would also solve some issues in gnembon/fabric-carpet#602

Reply:
If scripts want to use this, it's easier to just copy the code. so It's better to put it in a library where it is normal to find such code, as opposed to a script which does a bunch of other things as well.

Reply:
So in short, I would suggest extracting the item_consumed() func to some library, and importing form there, or signaling ot there, whatever.

Reply:
So are you suggesting adding a new .sc file to listen for an item being consumed and letting other scripts import that?

Reply:
imo just forget about the whole importing business, just add the function and be done with it.

Reply:
I'm not sure what you want me to do, the reason I'm signalling this as an event is so that other scripts that consume items can provide support for the persist option by signalling the event so the compass script can take care of replenishing the items if necessary. It's not really meant for other scripts to listen to.

Like for example, if someone was using the angel blocks script in a manhunt where the players spawn with cobblestone where the cobblestone never runs out, that wouldn't work when the player tries to place a block in the air, since the script removes stuff from the player's inventory, so to make it work, you could signal the event before removing the items, and the compass script will take care of checking if the persist option is used and replenishing the items if it is. I don't really think that importing a function would be the best way to do this.

If you're suggesting that I add a script that listens for when an item is used and signals an event, I'd be happy to do that, but it doesn't really seem that relevant to the compass script

Also if you're wondering why I didn't signal the event in the shoot_fireball script, it's because I don't have to, the compass script signals the event every time you use a fire charge, whether it's successful or not, I figured that just signalling the event anyways would be more efficient than doing ray tracing

Reply:
Yes, but you can't guarantee that someone has downloaded bot hthe compass AND shoot_fireball script.

Reply:
I'm sorry, but I'm extremely confused. It doesn't matter if people would have both scripts loaded, you can signal an event even if there's no script to listen for it. There's also no function that removes items, the function is meant to give people the option to replenish items whenever they're consumed so they never run out

Reply:
ye sorry I hadnt read code properly I meant that you should extract to a library

Reply:
Ok, I've added a new script that makes signalling the event a proper function